nilearn/connectome/rena_clustering.py

```diff
@@ -0,0 +1,622 @@
+        sparse matrix representation of the weighted adjacency graph
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    n_clusters : int
+        The number of clusters to find.
+
+    threshold : float in the close interval [0, 1], optional (default 1e-7)
+        The treshold is setted to handle eccentricities.
+        In practice it is 1e-7.
+
+    Returns
+    -------
+    reduced_connectivity : a sparse matrix in COOrdinate format.
+
+    reduced_masked_data :  array like
+        2D data matrix.
+
+    labels : array like
+        It contains the clusters assignation.
+    """
+    # Nearest neighbor conenctivity
+    nn_connectivity = _nn_connectivity(connectivity, threshold)
+    n_features = connectivity.shape[0]
+    n_components = n_features - (nn_connectivity + nn_connectivity.T).nnz / 2
+
+    if n_components < n_clusters:
+        # remove edges so that the final number of clusters is not less than
+        # n_clusters (to achieve the desired number of clusters)
+        n_edges = n_features - n_clusters
+        nn_connectivity = (nn_connectivity + nn_connectivity.T)
+
+        i_idx, j_idx = nn_connectivity.nonzero()
+        edges = np.array([i_idx, j_idx])
+
+        # select n_edges to merge.
+        edge_mask = np.argsort(i_idx - j_idx)[:n_edges]
+        # Set weights to 1, and the connectivity matrix symetrical.
+        weight = np.ones(2 * n_edges)
+        edges = np.hstack([edges[:, edge_mask], edges[::-1, edge_mask]])
+
+        nn_connectivity = coo_matrix((weight, edges),
+                                     (n_features, n_features))
+
+    # Clustering step: getting the connected components of the nn matrix
+    n_components, labels = csgraph.connected_components(nn_connectivity)
+
+    # Reduction step: reduction by averaging
+    reduced_connectivity, reduced_masked_data = _reduce_data_and_connectivity(
+        labels, n_components, connectivity, masked_data, threshold)
+
+    return reduced_connectivity, reduced_masked_data, labels
+
+
+def recursive_neighbor_agglomeration(masker, masked_data, n_clusters,
+                                     n_iter=10, threshold=1e-7):
+    """Recursive neighbor agglomeration: it performs iteratively the nearest
+    neighbor grouping.
+
+    Parameters
+    ----------
+    masker : NiftiMasker instance
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    n_clusters : int
+        The number of clusters to find.
+
+    n_iter : int, optional (default 10)
+        Number of iterations.
+
+    threshold : float in the close interval [0, 1], optional (default 1e-7)
+        The treshold is setted to handle eccentricities.
+        In practice it is 1e-7.
+
+    Returns
+    -------
+    n_components : int
+        Number of clusters.
+
+    labels : array
+        Cluster assignation.
```

Comment:
I think that here it would be useful to add a reference to the paper, as in a reference section (see:
https://numpydoc.readthedocs.io/en/latest/format.html )